HIP/hip-336.md

```diff
@@ -0,0 +1,301 @@
+	repeated TokenApproval tokenApproval = 2;
+}
+```
+
+This function contains a list of hbar and/or token approval messages. Each of these messages will set an allowance of `amount` tokens or hbar that can be transferred from the owner's account to another account by the `spender` account. Each of the hbar/token approvals in the list do not need to refer to the same spender. The owner can establish different allowances for various spenders in a single transaction.
+
+The total number of approvals contained within the transaction body cannot exceed 100. This number includes both hbar and token approvals.
+
+The `amount` must be specified in terms of the `decimals` value for the token. If `amount` is 0 the transaction will remove the `spender`'s allowance from the owner's account, thereby freeing an allowance for future use.
+
+It is not an error if the total token amount in the owner's account is less than the specified allowance `amount`. If the `spender` tries to transfer more tokens than exists in the owner's account that transaction will fail.
+
+It is not an error if the token is paused and an allowance approval is submitted.
+
+The transaction must be signed by the owner's account and transactioon fees will also be paid by this account.
+
+Each account is limited to 100 allowances. This limit spans hbar, fungible and non-fungible token allowances.
+
+The number of allowances set on an account will increase the auto renewal fee for the account. Conversely, removing allowances will decrease the auto renewal fee for the account.
+
+In order to grant the `spender` access to *all* of the owner's instances of a particular NFT, the `approvedForAll` field in the `TokenApproval` message should be set to `true`. However, the `approvedForAll` field cannot be set if `tokenId` refers to a fungible token.
+
+If the `spender` has already been granted `approvedForAll` for an owner's NFT and another transaction granting `approvedForAll` is sent for the same NFT, the transaction will be accepted and fees charged but it will effectively be a no-op. Similarly if multiple transactions to remove `approvedForAll` on the same NFT are received the effect will be idempotent.
+
+The allowance approval transaction will fail under the following circumstances:
+
+- The `spender` account is the same as the owner's account.
+- Either the `spender` or owner's accounts are frozen.
+- The `amount` will exceed the maximum supply of the token.
+- The owner account does not own the specified token.
+- `amount` is negative.
+- The referenced token is a fungible token and `approvedForAll` is specified.
+
+### Add CryptoAdjustAllowance API
+
+A new Hedera API will be added under the Crypto Service called `CryptoAdjustAllowance`. This function is called by the token owner to modify an existing allowance for a spender account.
+
+```protobuf
+message CryptoAdjustAllowanceTransactionBody {
+	TokenID token = 1;
+	AccountID spender = 2;
+	int64 amount = 3;
+}
```

Comment:
The `CryptoApproveAllowanceTransactionBody` call distinguishes crypto from other token transfers. But this API has some special meaning for `token` being null. Should the two calls have similar semantics?